### DISCUSSION TITLE
How can I associate metrics and traces

### DISCUSSION THREAD

**Discussion by Woodrow1026** [SigNoz/signoz#1795] [Q&A]
I am using opentelemetry to collect metrics and traces, and I realized that metrics include [Exemplar](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#exemplar), which can be used to correlate traces. However, when I send the indicator data to sigNoz, I found that the labels no longer contain Exemplar, and I cannot directly jump to the tracking in the indicator ( P99 in Services can do this).

Is there any way to achieve an exemplar-like effect here please?

**Comment by pranay01** [comment]
@Woodrow1026 We don't have exemplars support yet but you can go from metrics to traces for latency charts for services. If you click on any points in the latency chart you should see an option to `View Traces` which you can click to see the relevant traces.

Does this serve your use case?

![view-traces](https://user-images.githubusercontent.com/504541/204959390-be592c73-8713-4fad-bc11-c3c3ac3f9bc1.png)

**Reply by Woodrow1026** [reply]
Unfortunately, not quite the same.
We want to count the number of times a particular call fails in our service and be able to find the context of those failed calls.
If it's not possible to use Exemplar, is there any suggestion here to achieve something similar?

**Reply by Woodrow1026** [reply]
@pranay01 I am wondering, in our scenario, whether it is possible to associate metrics with traces, by carrying spanId in metrics or similar

**Reply by pranay01** [reply]
I think looking for more context on failed calls looks like a use case for traces. You should go to traces page, filter based on status code ( which will give you which calls have failed), filter for a specific service and then look into the details of traces

A quick video
https://www.loom.com/share/3f74cd323ac644a8b581e026f3a058c7

**Reply by Woodrow1026** [reply]
@pranay01 Thanks for your answer, but it doesn't seem to quite accomplish my goal

 When the request volume is relatively large, traces may not be sampled (probability sampling), and the response of the filtering operation is usually slower than directly displaying metrics.

The ideal situation is that after defining the corresponding metrics, I can jump directly from the metrics to the failed trace and get the context I need. But this means that traces and metrics must be associated??

**Reply by ankitnayan** [reply]
@Woodrow1026 we can also enable switching directly to a trace by simply choosing any trace matching the applied filters and hence skipping the list of representing traces in between. We can also think about supporting exemplars. 
We want to understand more on Exemplars. 
1. How do you add traceID to exemplars? Is it manual? 
2. Also, how do you limit the number of exemplars (how do you choose which one will be added when there are millions matching the criteria)?

**Reply by Woodrow1026** [reply]
@ankitnayan I'm sorry for the slow response, but I hope the following information is enough to answer your question.
1.Exemplars is specified in [OTel Doc](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#exemplar), and its content includes 
- the value of the Measurement that was recorded by the API call.
- The time the API call was made to record a Measurement.
- The set of [Attributes](https://opentelemetry.io/docs/reference/specification/common/#attribute) associated with the Measurement not already included in a metric data point.
- The associated [trace id and span id](https://opentelemetry.io/docs/reference/specification/trace/api/#retrieving-the-traceid-and-spanid) of the active [Span within Context](https://opentelemetry.io/docs/reference/specification/trace/api/#determining-the-parent-span-from-a-context) of the Measurement at API call time.

2.OpenTelemetry distinguishes Metrics into asynchronous and synchronous in the SDK. In the practice of Java SDK, I found that synchronous Metrics will automatically carry Exemplars. According to the documentation, the SDK provides [ExemplarFilter](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#exemplarfilter) for filtering, but I haven't practiced it yet.
